### PR TITLE
`Block` and `Transform` reworks

### LINKED ISSUE #608: Revit: issue with displaymesh conversions inside generic family document
Noticed when trying to convert blocks with breps from rhino into revit, which does so in a generic model family document. These brep meshes convert fine when not sent as a block.

Breps that correctly fall back to display mesh conversion fail when being converted in the family context, where the tesselated shape builder fails to build the mesh:

[https://speckle.xyz/streams/5dfbeb49c9/commits/22e9c409cc](https://speckle.xyz/streams/5dfbeb49c9/commits/22e9c409cc)

Stack:
TessellatedShapeBuilder failed to build the requested geometry.
   at Autodesk.Revit.DB.TessellatedShapeBuilder.Build()
   at Objects.Converter.Revit.ConverterRevit.MeshToNative(Mesh mesh, TessellatedShapeBuilderTarget target, TessellatedShapeBuilderFallback fallback)

### PR DESCRIPTION
**Pull request by izzylys**
## Description

This is a rework of Blocks to open the door for support in applications that don't properly support them (eg Revit).

TLDR: `Transform` is now a speckle object rather than a list of doubles representing the transform matrix. It allows you to apply the transform to points and vectors. Speckle objects can now implement `ITransformable` which involves writing a `TransformTo()` method. Meshes and some `ICurves` have been implemented as `ITransformable` and the Revit connector now supports them. The Revit connector does not truly support the concept of blocks, so new Groups are created for each `BlockInstance` with the transform baked into the source geometry. Note that this is p slow and could use optimisation.

- Fixes #608 indirectly

image

---

A new class `Objects.Other.Transform` as well as the interfaces `ITransformable` and `ITransformable<T>` have been added. The `Transform` class holds the matrix at `Transform.value` and contains helper methods for transforming points and vectors. The `ITransformable` interfaces can be implemented for any object that is transformable. You'll need to write a `TransformTo` method that returns a *new* object with the transformed values. `ITransformable<T>` is preferred as the return type of the `TransformTo` will match the current type, but `ITransformable` should be used in cases like curves where you need to check if an object implements `ITransformable` without knowing the type argument `<T>`. 

The only change to existing/future connectors is that the flat list of doubles representing the transform is accessed from `BlockInstance.transform.value` rather than `BlockInstance.transform`. This is a **⚠ BREAKING CHANGE** as the type of `BlockInstance.transform` is no longer a `double [ ]` but a `Transform`.

AutoCAD & Rhino have been updated to use the new transform:

image

image

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How has this been tested?

- Manual Tests with the usual test streams

## Docs

- Will be updated ASAP

### CODE REVIEW

File: Objects/Objects/Geometry/Brep.cs [hunk 1]
```diff
@@ -306,6 +307,33 @@ internal void OnDeserialized(StreamingContext context)
...
+    /// </summary>
+    /// <param name="transform"></param>
+    /// <returns></returns>
+    public Brep TransformTo(Transform transform)
+    {
+      return new Brep
+      {
+        provenance = provenance,
+        units = units,
+        displayMesh = displayMesh,
```
Comment (clairekuang) [inline Objects/Objects/Geometry/Brep.cs:323]: Is there a reason why the displayMesh isn't transformed here as well?

File: Objects/Objects/Geometry/Brep.cs [hunk 2]
```diff
@@ -306,6 +307,33 @@ internal void OnDeserialized(StreamingContext context)
...
       }
     }
+
+    /// <summary>
+    /// Returns a new transformed Brep. Note that `displayMesh` is *NOT* transformed. If you fall back on
+    /// the `displayMesh`, make sure to call `TransformTo` on the mesh object.
+    /// </summary>
+    /// <param name="transform"></param>
+    /// <returns></returns>
+    public Brep TransformTo(Transform transform)
```
Comment (clairekuang) [inline Objects/Objects/Geometry/Brep.cs:317]: Should consider redesigning these methods to return bool (or some other way of indicating a successful operation), since anything that implements `ApplyToCurves` can result in invalid geo which will be difficult to troubleshoot later.

File: Objects/Objects/Geometry/Brep.cs (hunk 1 above)
Comment (izzylys) [reply Objects/Objects/Geometry/Brep.cs:323]: performance 🙃
it's really slow to transform large meshes

File: Objects/Objects/Geometry/Brep.cs (hunk 2 above)
Comment (izzylys) [reply Objects/Objects/Geometry/Brep.cs:317]: that's a fair point! could we make an `out bool result` value? i think it's quite friendly for the method itself to return the transformed obj

Comment (izzylys) [reply Objects/Objects/Geometry/Brep.cs]: good point - refactor pushed! have a looksie ^^

**Review by clairekuang — APPROVED** [review]
Looks good to me!
